File: youtube/apis/playlist_items.py
"""PlaylistItems API namespace.

Wraps the following YouTube endpoints:

* GET    /playlistItems — list items in a playlist
* POST   /playlistItems — add a video to a playlist
* PUT    /playlistItems — update a playlist item
* DELETE /playlistItems — remove an item from a playlist

Reference: https://developers.google.com/youtube/v3/docs/playlistItems
"""

from __future__ import annotations

from collections.abc import AsyncIterator, Sequence
from typing import Any

from youtube.apis.base import BaseAPI
from youtube.apis.params import join_ids, join_parts
from youtube.config import YOUTUBE_API_BASE_URL
from youtube.models.playlist_items import (
    PlaylistItem,
    PlaylistItemListResponse,
    PlaylistItemPart,
)

_BASE = YOUTUBE_API_BASE_URL


class PlaylistItemsAPI(BaseAPI):
    """Methods for managing items within YouTube playlists."""

    async def list(
        self,
        *,
        parts: Sequence[PlaylistItemPart],
        playlist_id: str | None = None,
        id: str | Sequence[str] | None = None,
        video_id: str | None = None,
        max_results: int = 5,
        page_token: str | None = None,
    ) -> PlaylistItemListResponse:
        """Return items from a playlist."""
        params: dict[str, Any] = {
            "part": join_parts(parts),
            "maxResults": max_results,
        }
        if playlist_id is not None:
            params["playlistId"] = playlist_id
        if id is not None:
            params["id"] = join_ids(id)
        if video_id is not None:
            params["videoId"] = video_id
        if page_token is not None:
            params["pageToken"] = page_token

        payload = await self._session.get(f"{_BASE}/playlistItems", params=params)
        return PlaylistItemListResponse.model_validate(payload)

    async def iter_playlist(
        self,
        playlist_id: str,
        *,
        parts: Sequence[PlaylistItemPart],
        max_results: int = 50,
    ) -> AsyncIterator[PlaylistItem]:
        """Async generator that pages through all items in a playlist."""
        page_token: str | None = None
        while True:
            page = await self.list(
                parts=parts,
                playlist_id=playlist_id,
                max_results=max_results,
                page_token=page_token,
            )
            for item in page.items:
                yield item
            if not page.next_page_token:
                break
            page_token = page.next_page_token

    async def insert(
        self,
        body: PlaylistItem,
        *,
        parts: Sequence[PlaylistItemPart] | None = None,
    ) -> PlaylistItem:
        """Add a video to a playlist.

        The ``body`` must include ``snippet.playlist_id`` and
        ``snippet.resource_id`` (with ``kind`` and ``video_id``).
        """
        if parts is None:
            parts = _infer_parts(body)

        params: dict[str, Any] = {"part": join_parts(parts)}
        payload = await self._session.post(
            f"{_BASE}/playlistItems",
            json=body.model_dump(by_alias=True, exclude_none=True),
            params=params,
        )
        return PlaylistItem.model_validate(payload)

    async def update(
        self,
        body: PlaylistItem,
        *,
        parts: Sequence[PlaylistItemPart] | None = None,
    ) -> PlaylistItem:
        """Update a playlist item (e.g. change position or note)."""
        if parts is None:
            parts = _infer_parts(body)

        params: dict[str, Any] = {"part": join_parts(parts)}
        payload = await self._session.put(
            f"{_BASE}/playlistItems",
            json=body.model_dump(by_alias=True, exclude_none=True),
            params=params,
        )
        return PlaylistItem.model_validate(payload)

    async def delete(self, *, id: str) -> None:
        """Remove an item from a playlist."""
        await self._session.delete(f"{_BASE}/playlistItems", params={"id": id})


def _infer_parts(body: PlaylistItem) -> list[PlaylistItemPart]:
    parts: list[PlaylistItemPart] = []
    if body.snippet is not None:
        parts.append(PlaylistItemPart.SNIPPET)
    if body.content_details is not None:
        parts.append(PlaylistItemPart.CONTENT_DETAILS)
    return parts or [PlaylistItemPart.SNIPPET]

File: go/playlist_items.go
package youtube

import (
	"context"
	"errors"
	"net/http"
	"net/url"
	"strconv"
	"strings"
)

// PlaylistItem parts accepted by the part parameter.
const (
	PlaylistItemPartID             = "id"
	PlaylistItemPartSnippet        = "snippet"
	PlaylistItemPartContentDetails = "contentDetails"
	PlaylistItemPartStatus         = "status"
)

// PlaylistItemSnippet is the snippet part of a PlaylistItem resource.
type PlaylistItemSnippet struct {
	PublishedAt            string            `json:"publishedAt,omitempty"`
	ChannelID              string            `json:"channelId,omitempty"`
	Title                  string            `json:"title,omitempty"`
	Description            string            `json:"description,omitempty"`
	Thumbnails             *ThumbnailDetails `json:"thumbnails,omitempty"`
	ChannelTitle           string            `json:"channelTitle,omitempty"`
	PlaylistID             string            `json:"playlistId,omitempty"`
	Position               *int              `json:"position,omitempty"`
	ResourceID             *ResourceID       `json:"resourceId,omitempty"`
	VideoOwnerChannelTitle string            `json:"videoOwnerChannelTitle,omitempty"`
	VideoOwnerChannelID    string            `json:"videoOwnerChannelId,omitempty"`
}

// PlaylistItemContentDetails is the contentDetails part of a PlaylistItem.
type PlaylistItemContentDetails struct {
	VideoID          string `json:"videoId,omitempty"`
	Note             string `json:"note,omitempty"`
	VideoPublishedAt string `json:"videoPublishedAt,omitempty"`
}

// PlaylistItemStatus is the status part of a PlaylistItem resource.
type PlaylistItemStatus struct {
	PrivacyStatus string `json:"privacyStatus,omitempty"`
}

// PlaylistItem is a YouTube PlaylistItem resource (request body and response).
type PlaylistItem struct {
	Kind           string                      `json:"kind,omitempty"`
	Etag           string                      `json:"etag,omitempty"`
	ID             string                      `json:"id,omitempty"`
	Snippet        *PlaylistItemSnippet        `json:"snippet,omitempty"`
	ContentDetails *PlaylistItemContentDetails `json:"contentDetails,omitempty"`
	Status         *PlaylistItemStatus         `json:"status,omitempty"`
}

// PlaylistItemListResponse is the response from ListPlaylistItems.
type PlaylistItemListResponse struct {
	Kind          string         `json:"kind,omitempty"`
	Etag          string         `json:"etag,omitempty"`
	NextPageToken string         `json:"nextPageToken,omitempty"`
	PrevPageToken string         `json:"prevPageToken,omitempty"`
	PageInfo      *PageInfo      `json:"pageInfo,omitempty"`
	Items         []PlaylistItem `json:"items"`
}

// ListPlaylistItemsParams configures ListPlaylistItems.
type ListPlaylistItemsParams struct {
	// Parts is the list of resource parts to include. Required.
	Parts []string
	// PlaylistID returns items from the given playlist.
	PlaylistID string
	// IDs restricts the result to specific playlist-item IDs.
	IDs []string
	// VideoID restricts the result to items referencing the given video.
	VideoID string
	// MaxResults is the page size. Defaults to 5.
	MaxResults int
	// PageToken is the pagination cursor.
	PageToken string
}

// ListPlaylistItems returns items from a playlist. Mirrors playlistItems.list.
func (c *Client) ListPlaylistItems(ctx context.Context, p ListPlaylistItemsParams) (*PlaylistItemListResponse, error) {
	if len(p.Parts) == 0 {
		return nil, errors.New("youtube: ListPlaylistItems requires at least one part")
	}
	maxResults := p.MaxResults
	if maxResults <= 0 {
		maxResults = 5
	}
	q := url.Values{}
	q.Set("part", strings.Join(p.Parts, ","))
	q.Set("maxResults", strconv.Itoa(maxResults))
	if p.PlaylistID != "" {
		q.Set("playlistId", p.PlaylistID)
	}
	if len(p.IDs) > 0 {
		q.Set("id", strings.Join(p.IDs, ","))
	}
	if p.VideoID != "" {
		q.Set("videoId", p.VideoID)
	}
	if p.PageToken != "" {
		q.Set("pageToken", p.PageToken)
	}

	out := &PlaylistItemListResponse{}
	if err := c.doGet(ctx, "playlistItems", q, out); err != nil {
		return nil, err
	}
	return out, nil
}

// IterPlaylistItems walks every item in a playlist, invoking fn for each. It
// pages through the result set until exhausted or fn returns an error. This is
// the Go-idiomatic equivalent of the Python iter_playlist async generator.
func (c *Client) IterPlaylistItems(
	ctx context.Context,
	playlistID string,
	parts []string,
	maxResults int,
	fn func(PlaylistItem) error,
) error {
	if playlistID == "" {
		return errors.New("youtube: IterPlaylistItems requires a playlistID")
	}
	if maxResults <= 0 {
		maxResults = 50
	}
	pageToken := ""
	for {
		page, err := c.ListPlaylistItems(ctx, ListPlaylistItemsParams{
			Parts:      parts,
			PlaylistID: playlistID,
			MaxResults: maxResults,
			PageToken:  pageToken,
		})
		if err != nil {
			return err
		}
		for _, item := range page.Items {
			if err := fn(item); err != nil {
				return err
			}
		}
		if page.NextPageToken == "" {
			return nil
		}
		pageToken = page.NextPageToken
	}
}

// InsertPlaylistItem adds a video to a playlist. The body must include
// snippet.playlistId and snippet.resourceId (with kind and videoId). When parts
// is empty, the parts are inferred from the populated fields of the body.
// Mirrors playlistItems.insert. Requires OAuth.
func (c *Client) InsertPlaylistItem(ctx context.Context, body PlaylistItem, parts []string) (*PlaylistItem, error) {
	if len(parts) == 0 {
		parts = inferPlaylistItemParts(body)
	}
	q := url.Values{}
	q.Set("part", strings.Join(parts, ","))

	out := &PlaylistItem{}
	if err := c.doJSON(ctx, http.MethodPost, "playlistItems", q, body, out); err != nil {
		return nil, err
	}
	return out, nil
}

// UpdatePlaylistItem updates a playlist item (e.g. change position or note).
// When parts is empty, the parts are inferred from the populated fields of the
// body. Mirrors playlistItems.update. Requires OAuth.
func (c *Client) UpdatePlaylistItem(ctx context.Context, body PlaylistItem, parts []string) (*PlaylistItem, error) {
	if len(parts) == 0 {
		parts = inferPlaylistItemParts(body)
	}
	q := url.Values{}
	q.Set("part", strings.Join(parts, ","))

	out := &PlaylistItem{}
	if err := c.doJSON(ctx, http.MethodPut, "playlistItems", q, body, out); err != nil {
		return nil, err
	}
	return out, nil
}

// DeletePlaylistItem removes an item from a playlist by ID. Mirrors
// playlistItems.delete. Requires OAuth.
func (c *Client) DeletePlaylistItem(ctx context.Context, id string) error {
	if id == "" {
		return errors.New("youtube: DeletePlaylistItem requires an ID")
	}
	q := url.Values{}
	q.Set("id", id)
	return c.doDelete(ctx, "playlistItems", q)
}

// inferPlaylistItemParts derives the resource parts from the populated fields of
// a PlaylistItem body, mirroring the Python _infer_parts helper.
func inferPlaylistItemParts(body PlaylistItem) []string {
	parts := make([]string, 0, 2)
	if body.Snippet != nil {
		parts = append(parts, PlaylistItemPartSnippet)
	}
	if body.ContentDetails != nil {
		parts = append(parts, PlaylistItemPartContentDetails)
	}
	if len(parts) == 0 {
		return []string{PlaylistItemPartSnippet}
	}
	return parts
}
